### PR TITLE
Emoji panel and clipboard history: locate correct items if controls are different due to recent changes

### LINKED ISSUE #10377: Emoji panel and clipboard history: wrong item is announced when selecting first item in May 2019 Update and later
Hi,

Follow-up to #9204 

### Steps to reproduce:
For this to work, add-ons must be disabled:

1. From May 2019 Update, open eoji panel (Windows+period).
2. Open clipboard history (Widows+V).

### Actual behavior:
For clipboard history, wrong item is announced, and for emoji panel, something else is announced. This is observed on some systems.

### Expected behavior:
First selected item is announced like in the pas.

#### NVDA installed/portable/running from source:
Installed

#### NVDA version:
2019.3 alpha.18915

#### Windows version:
Windows 10 Version 1909 (November 2019 Update/build 18363)

#### Name and version of other software in use when reproducing the issue:
None

#### Other information about your system:
None

#### Does the issue still occur after restarting your PC?
Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
Yes - same behavior in 2019.1 and 2019.2.1

### Technical information:
This is caused by traversal issue - NVDA locates wrong item in modern keyboard UIA tree. A fix is included in Windows 10 App Essentials, which will be sent as a PR shortly.

Thanks.

### PR DESCRIPTION
**Pull request by josephsl**
### Link to issue number:
Fixes #10377 

### Summary of the issue:
In May 2019 Update, wrong item is announced when emoji panel and clipboard history opens due to UIA tree changes.

### Description of how this pull request fixes the issue:
Depending on UIA automation ID, select the right child from emoji panel/clipboard history panel so the correct item can be announced when it opens.

### Testing performed:
Tested via Windows 10 App Essentials add-on.

### Known issues with pull request:
None

### Change log entry:
None

### CODE REVIEW

File: source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py [hunk 1]
```diff
@@ -68,35 +68,55 @@ def event_UIA_window_windowOpen(self, obj, nextHandler):
...
 		# #9104: different aspects of modern input panel are represented by automation iD's.
-		childAutomationID = obj.firstChild.UIAElement.cachedAutomationID
+		inputPanelAutomationID = inputPanel.UIAElement.cachedAutomationID
 		# Emoji panel for build 16299 and 17134.
 		# This event is properly raised in build 17134.
-		if winVersion.winVersion.build <= 17134 and childAutomationID in ("TEMPLATE_PART_ExpressiveInputFullViewFuntionBarItemControl", "TEMPLATE_PART_ExpressiveInputFullViewFuntionBarCloseButton"):
+		if (
+			winVersion.winVersion.build <= 17134
+			and inputPanelAutomationID in (
+				"TEMPLATE_PART_ExpressiveInputFullViewFuntionBarItemControl",
```
Comment (leonardder) [inline source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py]: Could you store these in a constant?

File: source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py [hunk 2]
```diff
@@ -68,35 +68,55 @@ def event_UIA_window_windowOpen(self, obj, nextHandler):
...
 		# Handle hardware keyboard suggestions.
 		# Treat it the same as CJK composition list - don't announce this if candidate announcement setting is off.
-		elif childAutomationID == "CandidateWindowControl" and config.conf["inputComposition"]["autoReportAllCandidates"]:
+		elif (
+			inputPanelAutomationID == "CandidateWindowControl"
+			and config.conf["inputComposition"]["autoReportAllCandidates"]
+		):
 			try:
-				self.event_UIA_elementSelected(obj.firstChild.firstChild.firstChild, nextHandler)
+				self.event_UIA_elementSelected(inputPanel.firstChild.firstChild, nextHandler)
```
Comment (leonardder) [inline source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py]: I'm curious what @michaelDCurran thinks here, but I'd prefer something like `eventHandler.executeEvent` instead. IN the current code, it is difficult to find out how the flow of events is.

File: source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py [hunk 3]
```diff
@@ -68,35 +68,55 @@ def event_UIA_window_windowOpen(self, obj, nextHandler):
...
+		elif inputPanelAutomationID == "TEMPLATE_PART_ExpressionGroupedFullView":
 			self._emojiPanelJustOpened = True
+			# #10377: on some systems, there is something else besides grouping controls,
+			# so another child control must be used.
+			emojisList = inputPanel.children[-2]
+			if emojisList.UIAElement.cachedAutomationID != "TEMPLATE_PART_Items_GridView":
+				emojisList = emojisList.previous
 			try:
-				self.event_UIA_elementSelected(obj.firstChild.children[-2].firstChild.firstChild, nextHandler)
+				self.event_UIA_elementSelected(emojisList.firstChild.firstChild, nextHandler)
```
Comment (leonardder) [inline source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py]: Same as above

File: source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py [hunk 4]
```diff
@@ -68,35 +68,55 @@ def event_UIA_window_windowOpen(self, obj, nextHandler):
...
 		# #9103: if clipboard is empty, a status message is displayed instead, and luckily it is located where clipboard data items can be found.
-		elif childAutomationID == "TEMPLATE_PART_ClipboardTitleBar":
-			self.event_UIA_elementSelected(obj.children[-2], nextHandler)
+		elif inputPanelAutomationID == "TEMPLATE_PART_ClipboardTitleBar":
+			# Under some cases, clipboard tip text isn't shown on screen,
+			# causing clipboard history title to be announced instead of most recently copied item.
+			clipboardHistory = obj.children[-2]
+			if clipboardHistory.UIAElement.cachedAutomationID == inputPanelAutomationID:
+				clipboardHistory = clipboardHistory.next
+			self.event_UIA_elementSelected(clipboardHistory, nextHandler)
```
Comment (leonardder) [inline source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py]: Same as above

**Review by LeonarddeR — CHANGES_REQUESTED** [review]

**Comment by josephsl** [comment]
Hi, I’ll experiment with executeEvent in the add-on today, and as for using isWin10, that’s something I’ve considered last weekend (and will do so). As for emoji panel constants in Version 1709 and 1803, yes, I’ll do so in a list. Thanks.

**Comment by josephsl** [comment]
Hi,

Deferred to after 2019.3 release. Thanks.

**Comment by josephsl** [comment]
Hi,

Now that 2019.3 stable version is out, let's get this PR going again. Note that this PR might be influenced by a UIA IME branch being worked on by @MichaelDCurran.

Thanks.

**Comment by josephsl** [comment]
Hi,

I'm looking into an alternative implementation that may future-proof this and could serve as a test bed for File Explorer status bar announcement issue (don't remember exactly which ticket it was). basically I'm thinking about locating a specific UIA element based on automation ID rather than traversing through the UIA tree. If it works (to be tested via Windows 10 App Essentials add-on), I'll open a successor PR to resolve this unless it is fine to continue working on this branch.

Thanks.

**Comment by josephsl** [comment]
Hi,

Actually, scratch the alternative - traversal is still required when reading clipboard announcement when clipboard history is empty. So traversal is the way to go for this one.

Thanks.

**Comment by josephsl** [comment]
Potentially blocked by #11214. Once that PR is included in master, I suggest merging this PR, as the follow-up described in the mentioned PR will apply to this work also.

Thanks.

**Comment by feerrenrut** [comment]
@josephsl let me know if I misunderstood please.

This PR is blocked requiring #11214 in order to be able traverse the emoji and clipboard history. Labeling as blocked.

**Comment by josephsl** [comment]
Hi,

Technically, #11214 addresses name change event in emoji panel. This PR addresses UIA element traversal, which is different from that PR (note: the above comment was corrected).

Thanks.

**Comment by josephsl** [comment]
Hi,

Since #11447 was merged, this pull request has become a bit complicated, as modern keyboard support must be revised extensively thanks to detection of UIA events for elements with no Automation Id property. Therefore a future PR will succeed this one - closing for now (I plan to integrate this PR into the succeeding PR).

Thanks.

**Pull request closed without merging**